core/src/main/java/org/apache/calcite/rel/rules/ExpandDisjuctionRule.java

```diff
@@ -0,0 +1,379 @@
+    }
+
+    private Map<RexTableInputRef.RelTableRef, RexNode> expand(RexNode condition) {
+      try {
+        this.currentCount = 0;
+        return expandDeep(condition);
+      } catch (OverLimitException e) {
+        return new HashMap<>();
+      }
+    }
+
+    /**
+     * Expand predicates recursively that can be pushed down to single table.
+     *
+     * @param condition   Predicate to be expanded
+     * @return  Additional predicates that can be pushed down for each table
+     */
+    private Map<RexTableInputRef.RelTableRef, RexNode> expandDeep(RexNode condition) {
+      incrementAndCheck();
+      Map<RexTableInputRef.RelTableRef, RexNode> additionalConditions = new HashMap<>();
+      ImmutableBitSet inputRefs = RelOptUtil.InputFinder.bits(condition);
+      if (inputRefs.isEmpty()) {
+        return additionalConditions;
+      }
+      RexTableInputRef.RelTableRef tableRef = inputRefsBelongOneTable(inputRefs);
+      // The condition already belongs to one table, return it directly
+      if (tableRef != null) {
+        additionalConditions.put(tableRef, condition);
+        return additionalConditions;
+      }
+
+      // Recursively expand the expression according to whether it is a conjunction
+      // or a disjunction. If it is neither a disjunction nor a conjunction, it cannot
+      // be expanded further and an empty Map is returned.
+      switch (condition.getKind()) {
+      case AND:
+        List<RexNode> andOperands = RexUtil.flattenAnd(((RexCall) condition).getOperands());
+        for (RexNode andOperand : andOperands) {
+          Map<RexTableInputRef.RelTableRef, RexNode> operandResult = expandDeep(andOperand);
+          mergeAnd(additionalConditions, operandResult);
+        }
+        break;
+      case OR:
+        List<RexNode> orOperands = RexUtil.flattenOr(((RexCall) condition).getOperands());
+        additionalConditions.putAll(expandDeep(orOperands.get(0)));
+        for (int i = 1; i < orOperands.size(); i++) {
+          Map<RexTableInputRef.RelTableRef, RexNode> operandResult =
+              expandDeep(orOperands.get(i));
+          mergeOr(additionalConditions, operandResult);
+        }
+        break;
+      default:
+        break;
+      }
+
+      return additionalConditions;
+    }
+
+    private RexTableInputRef.@Nullable RelTableRef inputRefsBelongOneTable(
+        ImmutableBitSet inputRefs) {
+      RexTableInputRef.RelTableRef tableRef = inputRefToTableRef.get(inputRefs.nth(0));
+      if (tableRef == null) {
+        return null;
+      }
+      for (int inputBit : inputRefs.asList()) {
+        RexTableInputRef.RelTableRef inputBitTableRef = inputRefToTableRef.get(inputBit);
+        if (inputBitTableRef == null || !inputBitTableRef.equals(tableRef)) {
+          return null;
+        }
+      }
+      return tableRef;
+    }
+
+    /**
+     * For additional predicates of each operand in conjuction, all of them should be retained and
```

Comment:
conjuNction (I saw Mihai spotted a similar mispelling for disjunction), can you do a massive grep on this?